File: Week_04/Practicum/Flatten a Multilevel Doubly Linked List.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

// first way
class Solution {
public:
    pair<Node*, Node*> flat(Node* node) {
        Node* head = node;
        Node* next = node->next;

        if (node->child) {
            node->child->prev = node;
            auto p = flat(node->child);
            node->child = nullptr;
            node->next = p.first;
            node = p.second;
        } 


        if (next) {
            auto p = flat(next);
            node->next = p.first;
            node->next->prev = node;
            node = p.second;
        } 

        return {head, node};
    }
    Node* flatten(Node* head) {
        if (!head) {
            return head;
        }
        return flat(head).first;
    }
};

// second way
class Solution2 {
    Node* flattenAndReturnLastNode(Node* node) {
        Node* curr = node;
        while (curr) {
            if (curr->child) {
                Node* last = flattenAndReturnLastNode(curr->child);
                Node* next = curr->next;
                curr->child->prev = curr;
                curr->next = curr->child;
                curr->child = nullptr;

                if (!next) {
                    return last;
                } else {
                    next->prev = last;
                    last->next = next;
                    curr = next;
                }
            } else {
                if (!curr->next) {
                    return curr;
                }
                curr = curr->next;
            }
        }
        
        return curr;
    }

public:
    Node* flatten(Node* head) {
        if (!head) {
            return head;
        }
        
        flattenAndReturnLastNode(head);
        return head;
    }
};

File: Week_04/Practicum/Python/Flatten a Multilevel Doubly Linked List.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:

    def merge(self, curr):
        child = curr.child

        while child.next is not None:
            child = child.next
        
        if curr.next is not None:
            child.next = curr.next
            curr.next.prev = child

        curr.next = curr.child
        curr.child.prev = curr
        curr.child = None

    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        curr = head

        while curr is not None:
            if curr.child is not None:
                self.merge(curr)
            curr = curr.next
        
        return head

        